libs/core/synchronization/include/hpx/synchronization/counting_semaphore.hpp

```diff
@@ -22,41 +22,379 @@
+    ///        \namedrequirement{CopyAssignable},
+    ///        or \namedrequirement{MoveAssignable}.
+    ///
+    /// \note \a counting_semaphore_var's \a try_acquire() can spuriously fail.
+    ///
+    /// \tparam Mutex Type of mutex
+    /// \tparam N     The initial value of the internal semaphore lock
+    ///               count.
+    ///
+    template <typename Mutex = hpx::spinlock, int N = 0>
+    class counting_semaphore_var
+    {
+    private:
+        using mutex_type = Mutex;
+
+    public:
+        ///
+        /// \brief Constructs an object of type
+        ///        \a hpx::counting_semaphore_value
+        ///        with the internal counter initialized to \a N.
+        ///
+        /// \param value The initial value of the internal semaphore lock
+        ///              count. Normally this value should be zero, values
+        ///              greater than zero are equivalent to the same number of
+        ///              signals pre-set, and negative values are equivalent to
+        ///              the same number of waits pre-set. Defaults to \a N
+        ///              (which in turn defaults to zero).
+        ///
+        explicit counting_semaphore_var(std::ptrdiff_t value = N);
+
+        counting_semaphore_var(counting_semaphore_var const&) = delete;
+        counting_semaphore_var& operator=(
+            counting_semaphore_var const&) = delete;
+
+        ///
+        /// \brief Wait for the semaphore to be signaled.
+        ///
+        /// \param count The value by which the internal lock count will be
+        ///              decremented. At the same time this is the minimum
+        ///              value of the lock count at which the thread is not
+        ///              yielded.
+        ///
+        void wait(std::ptrdiff_t count = 1);
+
+        ///
+        /// \brief Try to wait for the semaphore to be signaled
+        ///
+        /// \param count The value by which the internal lock count will be
+        ///              decremented. At the same time this is the minimum
+        ///              value of the lock count at which the thread is not
+        ///              yielded.
+        ///
+        /// \return \a try_wait returns true if the calling thread was able to
+        ///         acquire the requested amount of credits.
+        ///         \a try_wait returns false if not sufficient credits are
+        ///         available at this point in time.
+        ///
+        bool try_wait(std::ptrdiff_t count = 1);
+
+        ///
+        /// \brief Signal the semaphore.
+        ///
+        /// \param count The value by which the internal lock count will be
+        ///              incremented.
+        ///
+        void signal(std::ptrdiff_t count = 1);
+
+        ///
+        /// \brief Unblock all acquirers.
+        ///
+        /// \return std::ptrdiff_t internal lock count after the operation.
+        ///
+        std::ptrdiff_t signal_all();
```

Comment:
@hkaiser is this documentation correct or would you word it differently? Also, I have guessed return value description; is it right?

Reply:
Yes, all good. Thanks!